### DISCUSSION TITLE
Baddet poisoning attack detector.fit()

### DISCUSSION THREAD

**Discussion by anseyyyy** [Trusted-AI/adversarial-robustness-toolbox#2390] [Q&A]
In this notebook: https://github.com/Trusted-AI/adversarial-robustness-toolbox/blob/main/notebooks/poisoning_attack_bad_det.ipynb

The notebook runs correctly until the very last line, ending in the following error:

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[16], line 1
----> 1 detector.fit(x_poisoned, y_poisoned, nb_epochs=1)

File E:\sepate_tests_for_thesis\env\lib\site-packages\art\estimators\object_detection\pytorch_object_detector.py:425, in PyTorchObjectDetector.fit(self, x, y, batch_size, nb_epochs, drop_last, scheduler, **kwargs)
    422     raise ValueError("An optimizer is needed to train the model, but none for provided.")
    424 # Apply preprocessing and convert to tensors
--> 425 x_preprocessed, y_preprocessed = self._preprocess_and_convert_inputs(x=x, y=y, fit=True, no_grad=True)
    427 class ObjectDetectionDataset(Dataset):
    428     """
    429     Object detection dataset in PyTorch.
    430     """

File E:\sepate_tests_for_thesis\env\lib\site-packages\art\estimators\object_detection\pytorch_object_detector.py:217, in PyTorchObjectDetector._preprocess_and_convert_inputs(self, x, y, fit, no_grad)
    214 x_tensor, y_tensor = cast_inputs_to_pt(x, y)
    216 if not self.channels_first:
--> 217     x_tensor = torch.permute(x_tensor, (0, 3, 1, 2))
    218 x_tensor = x_tensor / norm_factor
    220 # Set gradients

TypeError: permute(): argument 'input' (position 1) must be Tensor, not list

I tried changing the input type to tensors and a numpy array, but it still fails.

Any ideas why this might be or workarounds for the issue?

**Comment by beat-buesser** [comment]
Hi @antilaanssi What is the type of `x_poisoned` in the last cell?

**Reply by anseyyyy** [reply]
It is a list of numpy arrays. It should be a numpy array of numpy arrays for detector.fit() to work. However making the list of numpy arrays to numpy array of numpy arrays gives the following error:
```

---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/run/nvme/job_20432149/tmp/ipykernel_2099131/1850239494.py in <cell line: 4>()
      2 
      3 print(type(x_poisoned[0]))
----> 4 np_array_of_arrays = np.array(x_poisoned)

ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.
```

If calling detector.fit() with just one numpy.ndarray from the list, I get the following error:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/run/nvme/job_20432149/tmp/ipykernel_2099131/2349275260.py in <cell line: 1>()
----> 1 detector.fit(x_poisoned[0], y_poisoned, nb_epochs=1)

/scratch/project_2008539/lib/python3.10/site-packages/art/estimators/object_detection/pytorch_object_detector.py in fit(self, x, y, batch_size, nb_epochs, drop_last, scheduler, **kwargs)
    423 
    424         # Apply preprocessing and convert to tensors
--> 425         x_preprocessed, y_preprocessed = self._preprocess_and_convert_inputs(x=x, y=y, fit=True, no_grad=True)
    426 
    427         class ObjectDetectionDataset(Dataset):

/scratch/project_2008539/lib/python3.10/site-packages/art/estimators/object_detection/pytorch_object_detector.py in _preprocess_and_convert_inputs(self, x, y, fit, no_grad)
    215 
    216             if not self.channels_first:
--> 217                 x_tensor = torch.permute(x_tensor, (0, 3, 1, 2))
    218             x_tensor = x_tensor / norm_factor
    219 

RuntimeError: number of dims don't match in permute
```
Any ideas?
